include/experimental/__p2630_bits/submdspan_mapping.hpp

```diff
@@ -93,25 +81,71 @@ construct_sub_strides(const SrcMapping &src_mapping,
+template <class IndexType, size_t SubRank, class IndexSequence,
+          class... SliceSpecifiers>
+struct deduce_layout_left_submapping;
+
+template <class IndexType, size_t SubRank, size_t... Idx,
+          class... SliceSpecifiers>
+struct deduce_layout_left_submapping<
+    IndexType, SubRank, std::index_sequence<Idx...>, SliceSpecifiers...> {
+
+  using CountRange = index_sequence_scan_impl<
+      0, (std::is_convertible_v<SliceSpecifiers, IndexType> ? 0 : 1)...>;
+  //__static_partial_sums<!std::is_convertible_v<SliceSpecifiers,
+  // IndexType>...>;
+  constexpr static int NumGaps =
+      (((Idx > 0 && CountRange::get(Idx) == 1 &&
+         std::is_convertible_v<SliceSpecifiers, IndexType>)
+            ? 1
+            : 0) +
+       ... + 0);
+
+  constexpr static bool layout_left_value =
+      // Use layout_left for rank 0
+      (SubRank == 0) ||
+      // Use layout_left for rank 1 if leftmost slice specifier is range like
+      ((SubRank == 1) &&
+       ((Idx > 0 || std::is_same_v<SliceSpecifiers, full_extent_t> ||
+         std::is_convertible_v<SliceSpecifiers,
+                               std::tuple<IndexType, IndexType>>)&&...)) ||
+      // Preserve if leftmost SubRank-1 slices are full_extent_t and
+      // the leftmost SubRank slices are ranges
+      (((Idx < SubRank - 1 && std::is_same_v<SliceSpecifiers, full_extent_t>) ||
+        (Idx == SubRank - 1 &&
+         (std::is_same_v<SliceSpecifiers, full_extent_t> ||
+          std::is_convertible_v<SliceSpecifiers,
+                                std::tuple<IndexType, IndexType>>)) ||
+        (Idx >= SubRank && CountRange::get(Idx) == SubRank)) &&
+       ...);
+
+  constexpr static bool layout_left_padded_value =
+      // Use layout_left_padded for rank 0
       (SubRank == 0) ||
+      // Use layout_left_padded for rank 1 if leftmost slice specifier is range
+      // like
+      ((SubRank == 1) &&
+       ((Idx > 0 || std::is_same_v<SliceSpecifiers, full_extent_t> ||
+         std::is_convertible_v<SliceSpecifiers,
+                               std::tuple<IndexType, IndexType>>)&&...)) ||
+      // layout_left_padded case for SubRank > 1
       (
-          // Slice specifiers up to subrank need to be full_extent_t - except
-          // for the last one which could also be tuple but not a strided index
-          // range slice specifiers after subrank are integrals
-          ((Idx > SubRank - 1) || // these are only integral slice specifiers
-           (std::is_same_v<SliceSpecifiers, full_extent_t>) ||
-           ((Idx == SubRank - 1) &&
-            std::is_convertible_v<SliceSpecifiers, std::tuple<size_t, size_t>>)) &&
+          // leftmost must be range
+          ((Idx == 0 &&
+            (std::is_same_v<SliceSpecifiers, full_extent_t> ||
+             std::is_convertible_v<SliceSpecifiers,
+                                   std::tuple<IndexType, IndexType>>)) ||
+           (Idx > 0 && Idx <= NumGaps && CountRange::get(Idx) == 1) ||
```

Comment:
This would be greatly simplified with `if constexpr` in an initialization function